extensions/amp-selector/1.0/amp-selector.js

```diff
@@ -0,0 +1,348 @@
+import {CSS} from '../../../build/amp-selector-1.0.css';
+import {Option, Selector} from './selector';
+import {PreactBaseElement} from '../../../src/preact/base-element';
+import {Services} from '../../../src/services';
+import {
+  closestAncestorElementBySelector,
+  toggleAttribute,
+} from '../../../src/dom';
+import {createCustomEvent} from '../../../src/event-helper';
+import {dev, userAssert} from '../../../src/log';
+import {dict, omit} from '../../../src/utils/object';
+import {isExperimentOn} from '../../../src/experiments';
+import {mod} from '../../../src/utils/math';
+import {toArray} from '../../../src/types';
+import {useLayoutEffect} from '../../../src/preact';
+
+/** @const {string} */
+const TAG = 'amp-selector';
+
+class AmpSelector extends PreactBaseElement {
+  /** @override */
+  init() {
+    const {element} = this;
+    const action = Services.actionServiceForDoc(this.element);
+
+    // TODO(wg-bento): This hack is in place to prevent doubly rendering.
+    // See https://github.com/ampproject/amp-react-prototype/issues/40.
+    let isExpectedMutation = false;
+    const handleSelect = (value, option, trust) => {
+      fireSelectEvent(this.win, action, element, option, value, trust);
+      isExpectedMutation = true;
+      this.mutateProps(dict({'value': value}));
+    };
+
+    this.registerAction('clear', () => {
+      isExpectedMutation = true;
+      this.mutateProps(dict({'value': []}));
+    });
+
+    this.registerAction('selectUp', (invocation) => {
+      const {args, trust} = invocation;
+      const delta = args && args['delta'] !== undefined ? -args['delta'] : -1;
+      const initialValue = /** @type {!Array<string>} */ (this.getProp(
+        'value',
+        []
+      ));
+      const {value, option} = select(delta, initialValue, options);
+      handleSelect(value, option, trust);
+    });
+
+    this.registerAction('selectDown', (invocation) => {
+      const {args, trust} = invocation;
+      const delta = args && args['delta'] !== undefined ? args['delta'] : 1;
+      const initialValue = /** @type {!Array<string>} */ (this.getProp(
+        'value',
+        []
+      ));
+      const {value, option} = select(delta, initialValue, options);
+      handleSelect(value, option, trust);
+    });
+
+    this.registerAction(
+      'toggle',
+      (invocation) => {
+        const {args, trust} = invocation;
+        const {'index': index, 'value': opt_select} = args;
+        const initialValue = /** @type {!Array<string>} */ (this.getProp(
+          'value',
+          []
+        ));
+        const toggleValue = toggle(index, initialValue, options, opt_select);
+        if (!toggleValue) {
+          return;
+        }
+        const {'value': value, 'option': option} = toggleValue;
+        handleSelect(value, option, trust);
+      },
+      ActionTrust.LOW
+    );
+
+    const mu = new MutationObserver(() => {
+      if (isExpectedMutation) {
+        isExpectedMutation = false;
```

Comment:
I'm worried that this will contain multiple mutations at the same time. Eg, we do an expected mutation, then an unexpected one occurs and they both get lumped into the same notification batch.

Reply:
Yes this is definitely something we will have to fine tune more. The "pending set" approach described in [this thread](https://github.com/ampproject/amp-react-prototype/issues/40) seems appropriate but perhaps out of the scope of this initial implementation. What do you think?

Reply:
Definitely a big concern for me. But I think it's definitely outside the scope of this pull request. @jridgewell and any ideas are definitely welcome on the subject. Please comment on https://github.com/ampproject/amp-react-prototype/issues/40.